compiler/GHC/Parser.y

```diff
@@ -2793,13 +2799,32 @@ aexp    :: { ECP }
                                                   :(fst $ unLoc $4)) }
         | 'do' stmtlist              { ECP $
                                         $2 >>= \ $2 ->
-                                        amms (mkHsDoPV (comb2 $1 $2) (mapLoc snd $2))
+                                        amms (mkHsDoPV (comb2 $1 $2)
+                                                       Nothing
+                                                       (mapLoc snd $2))
                                                (mj AnnDo $1:(fst $ unLoc $2)) }
         | 'mdo' stmtlist            {% runPV $2 >>= \ $2 ->
                                        fmap ecpFromExp $
                                        ams (L (comb2 $1 $2)
                                               (mkHsDo (MDoExpr Nothing) (snd $ unLoc $2)))
                                            (mj AnnMdo $1:(fst $ unLoc $2)) }
+        -- QualifiedDo.
+        | QDO  stmtlist              { ECP $
+                                        $2 >>= \ $2 ->
+                                        amms (mkHsDoPV (comb2 $1 $2)
+                                                       (Just $
+                                                        mkModuleNameFS $
+                                                        getQDO $1)
+                                                       (mapLoc snd $2))
+                                               (mj AnnDo $1:(fst $ unLoc $2)) }
+        | QMDO stmtlist            {% runPV $2 >>= \ $2 ->
+                                       fmap ecpFromExp $
+                                       ams (L (comb2 $1 $2)
+                                              (mkHsDo (MDoExpr $
+                                                        Just $ mkModuleNameFS $
+                                                        getQMDO $1)
+                                                        (snd $ unLoc $2)))
+                                           (mj AnnMdo $1:(fst $ unLoc $2)) }
```

Comment:
Could we unify the `QDO` with the `do` case, and `QMDO` with `mdo`?

`getQDO` and `getQMDO` would yield `Maybe FastString` instead of `FastString`.

Reply:
This would also mean modifying the existing tokens `ITdo (Maybe FastString)` and `ITmdo (Maybe FastString)`, instead of introducing new tokens.

Reply:
I'll see if I can, though I suspect that would result in a reduce/reduce conflict.